src/main.rs

```diff
@@ -1,238 +1,400 @@
+        dimensions: [0.0, 0.0],
         depth_range: 0.0..1.0,
     };
 
-    let pipeline = get_pipeline(
-        device.clone(),
-        vs.clone(),
-        fs.clone(),
-        render_pass.clone(),
-        viewport.clone(),
-    );
-
-    let mut command_buffers =
-        get_command_buffers(&device, &queue, &pipeline, &framebuffers, &vertex_buffer);
+    let mut attachment_image_views = window_size_dependent_setup(&images, &mut viewport);
 
-    // main()
+    let command_buffer_allocator =
+        StandardCommandBufferAllocator::new(device.clone(), Default::default());
 
-    // Create infinity loop
-    let mut window_resized = false;
+    // swapchain can become invalid due to reasons like window losing focus, or resizing etc
     let mut recreate_swapchain = false;
 
-    event_loop.run(move |event, _, control_flow| match event {
-        Event::WindowEvent {
-            event: WindowEvent::CloseRequested,
-            ..
-        } => {
-            *control_flow = ControlFlow::Exit;
-        }
-        Event::WindowEvent {
-            event: WindowEvent::Resized(_),
-            ..
-        } => {
-            window_resized = true;
-        }
-        Event::MainEventsCleared => {}
-        Event::RedrawEventsCleared => {
-            if window_resized || recreate_swapchain {
-                recreate_swapchain = false;
-
-                let new_dimensions = surface.window().inner_size();
-
-                let (new_swapchain, new_images) = match swapchain.recreate(SwapchainCreateInfo {
-                    image_extent: new_dimensions.into(),
-                    ..swapchain.create_info()
-                }) {
-                    Ok(r) => r,
-                    Err(SwapchainCreationError::ImageExtentNotSupported { .. }) => return,
-                    Err(e) => panic!("Failed to recreate swapchain: {:?}", e),
-                };
-                swapchain = new_swapchain;
-                let new_framebuffers = get_framebuffers(&new_images, &render_pass);
-
-                if window_resized {
-                    window_resized = false;
-
-                    viewport.dimensions = new_dimensions.into();
-                    let new_pipeline = get_pipeline(
-                        device.clone(),
-                        vs.clone(),
-                        fs.clone(),
-                        render_pass.clone(),
-                        viewport.clone(),
-                    );
-                    command_buffers = get_command_buffers(
-                        &device,
-                        &queue,
-                        &new_pipeline,
-                        &new_framebuffers,
-                        &vertex_buffer,
-                    );
+    // Destroying the `GpuFuture` blocks until the GPU is finished executing it. In order to avoid
+    // that, we store the submission of the previous frame here.
+    let mut previous_frame_end = Some(sync::now(device.clone()).boxed());
```

Comment:
Not sure I follow, could you explain this to me like I'm 3, how does this allow us to execute the GPU code in a non-blocking way on line 354 compared to if we would not have this reference?